File: server/tasks/irac.py
#!/usr/bin/python2.7
# -*- encoding: utf8 -*-

"""
    Copyright (C) 2012-2015 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

import pythonlib.rnc_web as ws
from cc_modules.cc_html import tr_qa
from cc_modules.cc_string import WSTRING
from cc_modules.cc_task import (
    get_from_dict,
    STANDARD_TASK_FIELDSPECS,
    Task,
)


# =============================================================================
# IRAC
# =============================================================================

class Irac(Task):
    TASK_FIELDSPECS = [
        dict(name="aim", cctype="TEXT",
             comment="Main aim of the contact"),
        dict(name="achieved", cctype="INT", min=0, max=2,
             comment="Was the aim achieved? (0 not, 1 partially, 2 fully)"),
    ]
    TASK_FIELDS = [x["name"] for x in TASK_FIELDSPECS]

    @classmethod
    def get_tablename(cls):
        return "irac"

    @classmethod
    def get_taskshortname(cls):
        return "IRAC"

    @classmethod
    def get_tasklongname(cls):
        return "Identify and Rate the Aim of the Contact"

    @classmethod
    def get_fieldspecs(cls):
        return STANDARD_TASK_FIELDSPECS + cls.TASK_FIELDSPECS

    def is_complete(self):
        return (self.are_all_fields_complete(self.TASK_FIELDS)
                and self.field_contents_valid())

    def get_achieved_text(self):
        achieveddict = {
            None: None,
            0: WSTRING("irac_achieved_0"),
            1: WSTRING("irac_achieved_1"),
            2: WSTRING("irac_achieved_2"),
        }
        return get_from_dict(achieveddict, self.achieved)

    def get_task_html(self):
        if self.achieved is not None:
            achieved = u"{}. {}".format(self.achieved,
                                        self.get_achieved_text())
        else:
            achieved = None
        h = u"""
            <div class="summary">
                <table class="summary">
        """ + self.get_is_complete_tr() + u"""
                </table>
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="50%">Question</th>
                    <th width="50%">Answer</th>
                </tr>
        """
        h += tr_qa(WSTRING("irac_q_aim"), ws.webify(self.aim))
        h += tr_qa(WSTRING("irac_q_achieved"), achieved)
        h += u"""
            </table>
        """
        return h

File: tablet/Resources/task/Irac.js
// IRAC.js

/*
    Copyright (C) 2012-2015 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    // TABLE
    tablename = "irac",
    fieldlist = dbcommon.standardTaskFields();

fieldlist.push(
    {name: 'aim', type: DBCONSTANTS.TYPE_TEXT},
    {name: 'achieved', type: DBCONSTANTS.TYPE_INTEGER}
);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function IRAC(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(IRAC, taskcommon.BaseTask);
lang.extendPrototype(IRAC, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: IRAC,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Standard task functions
    isComplete: function () {
        return (this.aim !== null && this.achieved !== null);
    },

    getSummary: function () {
        return (
            L('irac_q_aim') + " " + this.aim +
            " -- " +
            L('irac_q_achieved') + " " + this.getAchievedText() +
            this.isCompleteSuffix()
        );
    },

    getAchievedText: function () {
        if (this.achieved === null || this.achieved < 0 || this.achieved > 2) {
            return "?";
        }
        return L('irac_achieved_' + this.achieved);
    },

    getDetail: function () {
        return this.getSummary();
    },

    edit: function (readOnly) {
        var self = this,
            KeyValuePair = require('lib/KeyValuePair'),
            Questionnaire = require('questionnaire/Questionnaire'),
            options_aim = [],
            options_achieved = [],
            i,
            pages,
            questionnaire,
            s;

        for (i = 1; i <= 10; ++i) {
            s = L("irac_aim_" + i);
            options_aim.push(new KeyValuePair(s, s));
        }
        for (i = 0; i <= 2; ++i) {
            options_achieved.push(new KeyValuePair(L("irac_achieved_" + i),
                                                   i));
        }

        pages = [{
            title: L("t_irac"),
            clinician: false,
            elements: [
                {
                    type: "QuestionText",
                    bold: true,
                    text: L("irac_q_aim"),
                },
                {
                    type: "QuestionMCQ",
                    field: "aim",
                    mandatory: true,
                    showInstruction: false,
                    horizontal: true,
                    asTextButton: true,
                    options: options_aim,
                },
                {
                    type: "QuestionText",
                    bold: true,
                    text: L("irac_q_achieved"),
                },
                {
                    type: "QuestionMCQ",
                    field: "achieved",
                    mandatory: true,
                    showInstruction: false,
                    horizontal: true,
                    asTextButton: true,
                    options: options_achieved,
                },
            ],
        }];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn,
        });
        questionnaire.open();
    },

});

module.exports = IRAC;
